components/canvas/canvas_data.rs

```diff
@@ -617,6 +618,89 @@ impl<'a> CanvasData<'a> {
         self.fill_text_with_size(text, x, y, max_width, is_rtl, size.px() as f64);
     }
 
+    /// <https://html.spec.whatwg.org/multipage/#text-preparation-algorithm>
+    /// <https://html.spec.whatwg.org/multipage/#dom-context-2d-measuretext>
+    pub fn measure_text(&mut self, text: String) -> TextMetrics {
+        // Step 2.
+        let text = replace_ascii_whitespace(text);
+        let font_style = self.state.font_style.as_ref();
+
+        let font = match font_style {
+            Some(style) => {
+                let font_group = self.font_context.font_group(style.clone());
+                let font = font_group
+                    .write()
+                    .first(&self.font_context)
+                    .expect("couldn't find font");
+                font
+            },
+            None => return TextMetrics::default(),
+        };
```

Comment:
For metrics that are taken from a single font, it sort of makes sense that you would use the first available font. When computing something like the advance of the text run though, I would expect to do measurement with the actual metrics of the rendered run (otherwise how could this API be useful?). Without doing the fallback the advance will be inaccurate.